File: tutorials/benchmark/steps.R
# Step recorder for the R side of the benchmark — the mirror of steps.py.
#
# Writes one JSON line per timed step. `as.numeric(Sys.time())` is epoch
# seconds, the same clock Python's `time.time()` reads, so the parent can align
# these boundaries with the RSS samples it took from outside the process.
#
# Deliberately hand-rolled JSON rather than jsonlite::toJSON: the log has to
# survive being written from a process that is about to run out of memory, and
# one `cat` per line with an open connection is the least that can go wrong.

step_log_open <- function(path) {
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  env <- new.env(parent = emptyenv())
  env$fh <- file(path, open = "w")
  env
}

.step_emit <- function(log, name, t0, t1, anchor) {
  anchor_json <- if (is.null(anchor)) "null" else {
    if (is.character(anchor)) sprintf('"%s"', anchor) else
      format(anchor, scientific = FALSE, trim = TRUE)
  }
  cat(sprintf('{"step": "%s", "t0": %.6f, "t1": %.6f, "seconds": %.6f, "anchor": %s}\n',
              name, t0, t1, t1 - t0, anchor_json),
      file = log$fh)
  flush(log$fh)
}

# Time `expr`. The expression's value becomes the step's anchor if it is a
# single number or string, and is returned to the caller either way.
step <- function(log, name, expr) {
  t0 <- as.numeric(Sys.time())
  value <- force(expr)
  t1 <- as.numeric(Sys.time())
  anchor <- if (is.atomic(value) && length(value) == 1L &&
                (is.numeric(value) || is.character(value))) value else NULL
  .step_emit(log, name, t0, t1, anchor)
  invisible(value)
}

# Record a step timed by hand (library() calls, mostly).
step_mark <- function(log, name, seconds, anchor = NULL) {
  t1 <- as.numeric(Sys.time())
  .step_emit(log, name, t1 - seconds, t1, anchor)
}

step_log_close <- function(log) close(log$fh)

File: tutorials/benchmark/steps.py
"""Step recorder for the Python side of the benchmark.

The child process writes one JSON line per timed step to a file the parent
(:mod:`run_benchmarks`) reads afterwards. Timestamps are epoch seconds from
:func:`time.time`, which is the same clock ``as.numeric(Sys.time())`` reads in
R — that is what lets the parent line one language's step boundaries up against
RSS samples it took of the other.

Each step also carries an ``anchor``: some cheap scalar summarising what the
step produced (cells kept, clusters found, markers returned). Anchors are not
timing data. They exist so the report can show the two arms did the *same*
work before it claims one of them did it faster.
"""
from __future__ import annotations

import json
import time
from contextlib import contextmanager
from pathlib import Path


class StepLog:
    """Append-only JSONL log of timed steps."""

    def __init__(self, path: str | Path) -> None:
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._fh = self.path.open("w")
        self._t_origin = time.time()

    @contextmanager
    def step(self, name: str):
        """Time a block, recording whatever the block puts in ``anchor``."""
        rec: dict = {}
        t0 = time.time()
        try:
            yield rec
        finally:
            t1 = time.time()
            self._emit({"step": name, "t0": t0, "t1": t1,
                        "seconds": t1 - t0, "anchor": rec.get("anchor")})

    def mark(self, name: str, seconds: float = 0.0, anchor=None) -> None:
        """Record a step that was timed by hand (an import, say)."""
        t1 = time.time()
        self._emit({"step": name, "t0": t1 - seconds, "t1": t1,
                    "seconds": seconds, "anchor": anchor})

    def _emit(self, rec: dict) -> None:
        self._fh.write(json.dumps(rec, default=str) + "\n")
        self._fh.flush()

    def close(self) -> None:
        self._fh.close()
